### PR TITLE
merge isAsciiWhitespace implementations

### PR DESCRIPTION
**Pull request by anonrig**
Merges 2 implementations and introduces 2 new cc_library targets on bazel.

### CODE REVIEW

File: src/workerd/api/BUILD.bazel [hunk 1]
```diff
@@ -38,6 +44,8 @@ api_header_exclusions = [
     "html-rewriter.h",
     "deferred-proxy.h",
     "data-url.h",
+    "encoding.h",
+    "util.h",
     "modules.h",
```
Comment (fhanau) [inline src/workerd/api/BUILD.bazel:49]: If a source file is being excluded somewhere, the corresponding header should also be excluded from the target, otherwise it's easy to get linker errors due to missing symbols (e.g. when depending on io but not on util-iocontext). I think the goal of splitting util.c++ to move it out of the io target is worth pursing, but either util-iocontext.c++ must be a part of the io target; or the one function it has should be moved into io-context.c++/.h itself (second option is cleaner IMHO).

File: src/workerd/api/BUILD.bazel [hunk 2]
```diff
@@ -159,6 +205,8 @@ wd_cc_capnp_library(
     kj_test(
         src = f,
         deps = [
+            ":encoding",
+            ":util-iocontext",
```
Comment (fhanau) [inline src/workerd/api/BUILD.bazel:209]: These tests need to depend on util-iocontext even though none of them directly use it, that could be avoided based on what I commented above

File: src/workerd/api/BUILD.bazel [hunk 3]
```diff
@@ -121,10 +129,48 @@ kj_test(
     ],
 )
 
+wd_cc_library(
+    name = "util",
+    srcs = ["util.c++"],
+    hdrs = ["util.h"],
+    visibility = ["//visibility:public"],
+    deps = [
+        "//src/workerd/jsg",
```
Comment (fhanau) [inline src/workerd/api/BUILD.bazel:138]: Hope this isn't too much bikeshedding, just want to get this target right: @simdutf should be implementation_deps here since it's only used in the .c++ file

Comment (anonrig) [reply src/workerd/api/BUILD.bazel:138]: The comments and the reviews are extremely helpful. Thank you! (And I don't consider this as bike-shedding)

File: src/workerd/io/io-timers.h [hunk 4]
```diff
@@ -86,4 +86,7 @@ class TimeoutManager {
   virtual kj::Maybe<kj::Date> getNextTimeout() const = 0;
 };
 
+// Returns exactly what Date.now() would return.
```
Comment (fhanau) [inline src/workerd/io/io-timers.h:89]: This is not needed here right?

File: src/workerd/io/BUILD.bazel [hunk 5]
```diff
@@ -332,13 +335,17 @@ kj_test(
 
 kj_test(
     src = "promise-wrapper-test.c++",
-    deps = [":io"],
+    deps = [
+        ":io",
+        "//src/workerd/api:encoding",
```
Comment (fhanau) [inline src/workerd/io/BUILD.bazel:340]: Neither of these two tests depends directly on encoding right? Let's IWYU here and leave these as they were before

File: src/workerd/api/util.c++ [hunk 6]
```diff
@@ -259,15 +257,6 @@ kj::String redactUrl(kj::StringPtr url) {
   return kj::String(redacted.releaseAsArray());
 }
 
-double dateNow() {
```
Comment (jasnell) [inline src/workerd/api/util.c++:262]: Unrelated change?

File: src/workerd/api/BUILD.bazel [hunk 7]
```diff
@@ -177,6 +213,7 @@ kj_test(
     src = "data-url-test.c++",
     deps = [
         ":data-url",
+        ":encoding",
```
Comment (fhanau) [inline src/workerd/api/BUILD.bazel:216]: The test doesn't depend on encoding either. The same goes for api-rtti-test – encoding.h is included in there but not actually needed I think.

File: src/workerd/api/util.c++ (hunk 6 above)
Comment (jasnell) [reply src/workerd/api/util.c++:262]: ah, nevermind, I see that things were shifted around a bit.

File: src/workerd/api/data-url.c++ [hunk 8]
```diff
@@ -11,18 +13,6 @@ kj::Maybe<DataUrl> DataUrl::tryParse(kj::StringPtr url) {
...
 
-static constexpr kj::FixedArray<uint8_t, 256> ascii_whitespace_table = []() consteval {
-  kj::FixedArray<uint8_t, 256> result{};
-  for (uint8_t c: {0x09, 0x0a, 0x0c, 0x0d, 0x20}) {
-    result[c] = true;
-  }
-  return result;
-}();
-
-constexpr bool isAsciiWhitespace(uint8_t c) noexcept {
```
Comment (fhanau) [inline src/workerd/api/data-url.c++:22]: Conceptually, this could affect performance – `isAsciiWhitespace()` does a single table lookup so having it defined in a different source file will make using it much slower based on needing a function call now. The compiler may or may not be able to inline it properly if LTO is enabled (such as with the downstream release build), so this is non-blocking.

File: src/workerd/api/util.c++ (hunk 6 above)
Comment (anonrig) [reply src/workerd/api/util.c++:262]: yes, this change is required in order to avoid depending on io on this file, which eventually creates a circular dependency. the best way to avoid this by moving it to io-util.h

File: src/workerd/api/data-url.c++ (hunk 8 above)
Comment (anonrig) [reply src/workerd/api/data-url.c++:22]: I moved the implementation to encoding.h and made it constexpr again. Since it is an extremely small fixed array, I don't think it would have any negative impact, but let me know if you disagree. I can change it back without any hassle.

**Review by jasnell — APPROVED** [review]

**Comment by anonrig** [comment]
...just in case...

**Review by fhanau — APPROVED** [review]

**Review by mikea — APPROVED** [review]
